datalad/core/local/diff.py

```diff
@@ -0,0 +1,335 @@
+    rev_resolve_path,
+    path_under_rev_dataset,
+    rev_get_dataset_root,
+)
+
+from datalad.support.constraints import (
+    EnsureNone,
+    EnsureStr,
+)
+from datalad.support.param import Parameter
+from datalad.consts import PRE_INIT_COMMIT_SHA
+
+from datalad.core.local.status import (
+    Status,
+    _common_diffstatus_params,
+)
+
+lgr = logging.getLogger('datalad.core.local.diff')
+
+
+@build_doc
+class Diff(Interface):
+    """Report differences between two states of a dataset (hierarchy)
+
+    The two to-be-compared states are given via to --from and --to options.
+    These state identifiers are evaluated in the context of the (specified
+    or detected) dataset. In case of a recursive report on a dataset
+    hierarchy corresponding state pairs for any subdataset are determined
+    from the subdataset record in the respective superdataset. Only changes
+    recorded in a subdataset between these two states are reported, and so on.
+
+    Any paths given as additional arguments will be used to constrain the
+    difference report. As with Git's diff, it will not result in an error when
+    a path is specified that does not exist on the filesystem.
+
+    Reports are very similar to those of the `rev-status` command, with the
+    distinguished content types and states being identical.
+    """
+    # make the custom renderer the default one, as the global default renderer
+    # does not yield meaningful output for this command
+    result_renderer = 'tailored'
+
+    _params_ = dict(
+        _common_diffstatus_params,
+        path=Parameter(
+            args=("path",),
+            metavar="PATH",
+            doc="""path to contrain the report to""",
+            nargs="*",
+            constraints=EnsureStr() | EnsureNone()),
+        fr=Parameter(
+            args=("-f", "--from",),
+            dest='fr',
+            metavar="REVISION",
+            doc="""original state to compare to, as given by any identifier
+            that Git understands.""",
+            nargs=1,
+            constraints=EnsureStr()),
+        to=Parameter(
+            args=("-t", "--to",),
+            metavar="REVISION",
+            doc="""state to compare against the original state, as given by
+            any identifier that Git understands. If none is specified,
+            the state of the worktree will be used compared.""",
+            nargs=1,
+            constraints=EnsureStr() | EnsureNone()),
+    )
+
+    @staticmethod
+    @datasetmethod(name='diff')
+    @eval_results
+    def __call__(
+            path=None,
+            fr='HEAD',
+            to=None,
+            dataset=None,
+            annex=None,
+            untracked='normal',
+            recursive=False,
+            recursion_limit=None):
+        ds = require_dataset(
+            dataset, check_installed=True, purpose='difference reporting')
+
+        # convert cmdline args into plain labels
+        if isinstance(fr, list):
+            fr = fr[0]
```

Comment:
I guess just for my own comprehension -- why do we need `nargs=1` above which is what I guess possibly leads to a `list` here? (note that the `if` body isn't even triggered by the tests)

Reply:
I cannot recall, or construct a reason. Removed.